### PR TITLE
Use SvelteKit to build templates

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

This is a complete rewrite of how we build commercial templates. We have found it harder and harder to work on these templates, as they are seldom updated and it’s hard to understand what each template looks like in practice. The ever-growing amount of CSS makes it hard to make changes confidently. By using components and making the templates be deployed on their own microsite, we can solve these two issues.

Development of the templates is also improved, as templates can be continuously rebuilt whilst working on them.

In terms of technology, this uses Svelte components for building template elements. SvelteKit is how we have a preview environment, that we can publish on github pages as a static site.

- [X] Hot reloading
- [X] Deployment to GH Pages (not automated yet)
- [X] Real Dynamic (CSR) example (`capi-single-paidfor`)
- [x] Real Static (SSR) example (outstanding: include some JS)
- [ ] Typings of iframe messenger

All the the previous files are kept in `legacy`.

## How to test

`npm run dev`

## How can we measure success?

Easier to work with, live reloading, modern tooling.

The resulting code is no larger that the existing one, and performs as fast.

## Have we considered potential risks?

There’s a lot of knowledge that went into the original templates. It’s possible that new templates initially lack some of the features. This needs to be thoroughly tested before being put to production.

## Images

image

### CODE REVIEW

File: src/lib/messenger.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+type StringMessage = StandardMessage<'message', string>;
+
+type Message = ResizeMessage | StringMessage;
+
+const post = (arg: Message): void => {
+	window.top?.postMessage({ id: self.name, ...arg }, '/');
+};
+
+export { post };
+export type { Message };
```
Comment (mxdvl) [inline src/lib/messenger.ts:28]: @chrislomaxjones just for you 😉

File: README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Aiming to provide a great developer and reader experience, this project uses
+[Svelte][] to build the template, and [SvelteKit][] to help preview development.
+You can see the current templates on [guardian.github.io/commercial-templates](https://guardian.github.io/commercial-templates/).
+
+[svelte]: https://svelte.dev/
+[sveltekit]: https://kit.svelte.dev/
+
+This project is a rewrite and work-in-progress. Legacy templates can still be
+found in the [`/legacy` folder](/legacy)
```
Comment (mxdvl) [inline README.md:16]: Is this enough of an intro?

File: src/routes/csr/[template].json.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		`<div id="svelte" data-template-id="${template}"></div>`,
+		`<div id="metrics"></div><style>#metrics{position: fixed; top: 0; left:0; background: #0004; color: white;}</style>`,
+		`<script>${String(chunks[0].code)}</script>`,
+	].join('\n');
+
+	const css = [`/* ${stamp} */`, styles].join('\n');
+
+	const description = existsSync(`${dir}/README.md`)
+		? marked.parse(readFileSync(`${dir}/README.md`, 'utf-8'))
+		: `<p><em>no description provided</em></p>`;
```
Comment (mxdvl) [inline src/routes/csr/[template].json.ts:58]: This is not the most DRY… but I think it’s fine to have different build processes for CSR and SSR.

File: src/routes/ssr/[template].json.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	const ctx = vm.createContext({ setShare });
+	vm.runInContext(code, ctx);
+
+	return {
+		// @ts-expect-error -- it’s happening in the vm
+		html: share.html,
+		// @ts-expect-error -- it’s happening in the vm
+		css: share.css.code,
+	};
+};
```
Comment (mxdvl) [inline src/routes/ssr/[template].json.ts:38]: @arelra sorry not sorry 🎰

File: src/templates/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+There are two main types of templates, which live in different folders.
+
+Static templates are bare HTML + CSS, where Ad Manager variables are replaced by
+the create serving them. Many of the templates fit this category, such as `fabric`,
+`fabric-video`, `manual`, etc. They live in the [`src/templates/ssr/` folder](ssr/).
+
+Dynamic templates need to make a request to an api in order to generate their
+content. These are mainly the `capi-*` templates, e.g. `capi-single-paidfor`.
+These only generate JS + CSS, and the HTML is generated on the client by svelte.
+They live in the [`csr/` folder](csr/).
```
Comment (mxdvl) [inline src/templates/README.md:49]: @guardian/commercial-dev, this is worth reviewing!

File: src/templates/csr/capi-single-paidfor/index.svelte [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		background-color: #69d1ca;
+		color: white;
+	}
+
+	@media (min-width: 1140px) {
+		aside {
+			flex-direction: row;
+		}
+	}
+</style>
```
Comment (mxdvl) [inline src/templates/csr/capi-single-paidfor/index.svelte:67]: Dynamic / CSR example.

File: src/templates/ssr/manual-single/index.svelte [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		background: #ffe500;
+		color: black;
+		text-decoration: none;
+		border-radius: 10rem;
+		padding: 3px;
+		padding-left: 1rem;
+		align-self: flex-start;
+		flex-shrink: 0;
+	}
+</style>
```
Comment (mxdvl) [inline src/templates/ssr/manual-single/index.svelte:147]: Static / SSR example.

File: svelte.config.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+						/** @type {import('./src/lib/reload').Data} */
+						const data = {
+							id: matches[1],
+						};
+
+						ctx.server.ws.send({
+							type: 'custom',
+							event: 'template-update',
+							data,
+						});
```
Comment (mxdvl) [inline svelte.config.js:54]: This allows invalidation of `ssr/[template].json` endpoints when source files are changed.

**Review by mxdvl — COMMENTED** [review]
Below are some of the key bits to review 👇

File: src/templates/README.md (hunk 5 above)
Comment (chrislomaxjones) [reply src/templates/README.md:49]: Reads clearly to me.

File: svelte.config.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import adapter from '@sveltejs/adapter-static';
+import preprocess from 'svelte-preprocess';
+import { defineConfig } from 'vite';
+
+const isDev = process.env.NODE_ENV === 'development';
+
+/** @type {import('@sveltejs/kit').Config} */
```
Comment (chrislomaxjones) [inline svelte.config.js:7]: TIL `/** @type ... */`

File: README.md (hunk 2 above)
Comment (chrislomaxjones) [reply README.md:16]: Seems like more than enough to get started with this repo

File: src/routes/ssr/[template].json.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+type Output = {
+	html?: string;
+	js?: string;
+	css?: string;
+};
+
+const github = 'https://github.com/guardian/commercial-templates/blob';
+
+const prerender = (code: string): Output => {
```
Comment (chrislomaxjones) [inline src/routes/ssr/[template].json.ts:18]: This function still remains a bit mystical to me, will have to spend some time looking over this when we start rebuilding more templates.

**Review by chrislomaxjones — APPROVED** [review]
Nice job @mxdvl, this is seriously impressive! A big improvement in terms of developer experience.

This is a big PR and I've endeavoured to review in detail the static/dynamic templates and READMEs. Still, I'm approving this since it's low-risk, in terms of not automatically deploying anything new and not invalidating any of our old templates, just marking them as legacy. 

I think next steps are for the team to get up to speed by actually trying to rebuild some the key templates.

File: src/lib/types/capi.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
+export type Source = {
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply src/lib/types/capi.ts:1]: Added some links for the `Branding` and `Single` types’ origin.

File: src/lib/types/capi.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+	minWidth: string;
+	sizes: string;
+	hidpiSrcset: string;
+	lodpiSrcset: string;
+};
+
+export type Branding = {
+	brandingType: {
+		name: string;
+	};
```
Comment (mxdvl) [inline src/lib/types/capi.ts:11]: Not allowed to commit suggestions any longer 😢 – can you, @chrislomaxjones ?

```suggestion
/**
 * Best effort Scala conversion from [Branding.scala](https://github.com/guardian/commercial-shared/blob/a692e8b2eba6e79eeeb666e5594f2193663f6514/src/main/scala/com/gu/commercial/branding/Branding.scala#L7-L22)
 */
export type Branding = {
	brandingType: {
		name: string;
	};
```

File: src/lib/types/capi.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+			width: number;
+			height: number;
+		};
+		link: string;
+		label: string;
+	};
+	edition: 'UK' | 'AU' | 'US';
+};
+
+export type Single = {
```
Comment (mxdvl) [inline src/lib/types/capi.ts:25]: Same as above…

```suggestion
/**
 * Best effort at interpreting Scala types.
 *
 * [Source in `frontend`](https://github.com/guardian/frontend/blob/9d2ee56372de4bb730146100527f83b2efd41c35/commercial/app/model/capi/CapiSingle.scala#L9-L19)
 *
 * [route conf](https://github.com/guardian/frontend/blob/9d2ee56372de4bb730146100527f83b2efd41c35/commercial/conf/routes#L30)
 */
export type Single = {
```

**Review by mxdvl — COMMENTED** [review]
Thanks a lot, can you please commit my two suggestions, @chrislomaxjones? I don’t have write rights any longer.

File: src/lib/types/capi.ts (hunk 12 above)
Comment (chrislomaxjones) [reply src/lib/types/capi.ts:14]: Of course!

**Comment by chrislomaxjones** [comment]
> Thanks a lot, can you please commit my two suggestions, @chrislomaxjones? I don’t have write rights any longer.

Have done! I also assume you can't merge, so will do that as well.

**Pull request merged**